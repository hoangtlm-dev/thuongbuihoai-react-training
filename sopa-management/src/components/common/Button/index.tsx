import React, { memo } from "react";
import "./button.css";

export enum ButtonType {
  default = "",
  primary = "primary",
  secondary = "secondary",
  tertiary = "tertiary",
  btnIconPrimary = "icon-primary",
  btnIconSecondary = "icon-secondary"
};

interface ButtonProps {
  children?: string | JSX.Element;
  type?: ButtonType;
  disable?: boolean;
  onClick?: () => void;
  onSubmit?: () => void;
};

const Button: React.FC<ButtonProps> = ({
  children = "",
  type = ButtonType.default,
  disable = false,
  onClick = () => {},
  onSubmit = () => {},
}) => {
  let className = "btn";
  switch (type) {
    case ButtonType.primary:
      className += ` btn-${ButtonType.primary}`;
      break;
    case ButtonType.secondary:
      className += ` btn-${ButtonType.secondary}`;
      break;
    case ButtonType.tertiary:
      className += ` btn-${ButtonType.tertiary}`;
      break;
    case ButtonType.btnIconPrimary:
      className += ` btn-${ButtonType.btnIconPrimary}`;
      break;
    case ButtonType.btnIconSecondary:
      className += ` btn-${ButtonType.btnIconSecondary}`;
      break;
    default:
      break;
  };

  return (
    <button
      data-testid="button"
      className={`${className} btn-${type}`}
      disabled={disable}
      onClick={onClick}
      onSubmit={onSubmit}
    >{children}</button>
  );
};

export default memo(Button);
