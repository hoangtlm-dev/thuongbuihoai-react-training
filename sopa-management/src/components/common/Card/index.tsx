import { memo } from 'react';
import { Link } from 'react-router-dom';

// Components
import Text,
{
  ThemeType
} from '@common/Text';
import Price,
{
  PriceType
} from '@common/Price';
import Image from '@common/Image';

// Styles
import './card.css';

interface CardProps {
  key?: string;
  name?: string;
  color?: string;
  price?: number;
  href: string;
  src: string;
  alt?: string;
}

const Card: React.FC<CardProps> = ({
  name = '',
  color = '',
  price = 0,
  href,
  src,
  alt = ''
}) => {
  return (
    <div data-testid='card' className='card'>
      <div className='card-item'>
        <figure className='card-image'>
          <Image href={href} url={src} alt={alt} className='card-image-item' />
        </figure>
        <div className='card-info'>
          <Link to={href} className='card-name'>
            {name}
          </Link>
          <div className='card-description'>
            <Text text={color} type={ThemeType.highlightPrimary} />
            <Price value={price} type={PriceType.primary} />
          </div>
        </div>
      </div>
    </div>
  );
};

export default memo(Card);
