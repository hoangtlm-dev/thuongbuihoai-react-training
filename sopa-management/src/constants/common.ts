const SIZE = [
  {
    key: "1",
    label: "Small",
  },
  {
    key: "2",
    label: "Medium",
  },
  {
    key: "3",
    label: "Large",
  }
];

const USER = [
  {
    key: "1",
    label: "Logout"
  },
];

export {
  SIZE,
  USER,
};
