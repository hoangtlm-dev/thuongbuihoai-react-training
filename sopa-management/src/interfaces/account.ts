export interface Account {
  id: string;
  email: string;
  password: string;
};

export interface LoginType {
  default: '',
  info: 'info',
  error: 'error'
};
