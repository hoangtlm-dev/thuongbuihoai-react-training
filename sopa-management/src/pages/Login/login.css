.form {
  margin: 80px 0 200px;
  text-align: center;
  height: 540px
}

.form-input {
  width: var(--w-extra-medium);
  margin: 0 auto;
}

.form-email {
  margin: 60px 0 24px;
}

.form-password {
  margin-bottom: 24px;
}

.form-error {
  text-align: left;
  color: var(--border-tertiary);
  font-size: var(--text-sm);
  margin-top: 3px;
  padding-left: var(--py-extra-base);
}
