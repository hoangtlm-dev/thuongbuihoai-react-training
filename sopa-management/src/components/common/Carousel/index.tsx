// Images
import { IMAGE } from '@constants/image';

// Components
import Button,
{
  ButtonType
} from '@common/Button';
import Image from '@common/Image';
import Input from '@common/Input';
import Text,
{
  SizeType
} from '@common/Text';

// Constants
import { CAROUSEL_BULLET } from '@constants/common';

// Styles
import './carousel.css';

const Carousel: React.FC = () => (
  <div data-testid='carousel' className='carousel'>
    <div className='carousel-inner'>
      <Input
        className='carousel-open'
        type='radio'
        id='carousel-1'
        name='carousel'
        hidden
        checked
      />
      <figure className='carousel-item'>
        <Image url={IMAGE.banner} alt={IMAGE.altBanner} />
      </figure>
      <Input
        className='carousel-open'
        type='radio'
        id='carousel-2'
        name='carousel'
        hidden
      />
      <figure className='carousel-item'>
        <Image url={IMAGE.banner2} alt={IMAGE.altBanner} />
      </figure>
      <Input
        className='carousel-open'
        type='radio'
        id='carousel-3'
        name='carousel'
        hidden
      />
      <figure className='carousel-item'>
        <Image url={IMAGE.banner3} alt={IMAGE.altBanner} />
      </figure>
      <ul className='carousel-indicators'>
        {CAROUSEL_BULLET.map((item) => (
          <li key={item.key}>
            <label htmlFor={item.htmlFor} className={item.className} />
          </li>
        ))}
      </ul>
    </div>
    <div className='carousel-info'>
      <Text
        text='Step inside, for comfort and magic await you.'
        type={SizeType.large}
        className='carousel-heading'
      />
      <Button children='Show now' type={ButtonType.primary} />
    </div>
  </div>
);

export default Carousel;
