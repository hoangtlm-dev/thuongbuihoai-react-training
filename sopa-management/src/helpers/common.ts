import { FieldError } from 'react-hook-form';

// Interfaces
import { Account } from '@interfaces/account';
import { Product } from '@interfaces/product';
import {
  InputTheme,
  InputType
} from '@components/common/Input';

// Flattens an array of arrays into a single array
export const flattenArray = <T>(pages: T[][]): T[] => {
  const result: T[] = [];
  pages.forEach((page) => {
    result.push(...page);
  });

  return result;
};

// Calculate the total price
export const totalPrices = (carts: Product[]) => {
  const subTotals = carts.map((item: Product) => item.quantity * item.price);

  const total = subTotals.reduce(
    (totalPrice: number, cartItem: number) => totalPrice + cartItem,
    0
  );

  return total;
};

// Check login
export const checkLogin = (
  data?: Account[],
  email?: string,
  password?: string
) => {
  return data?.some(
    (user: Account) => user.email === email && user.password === password
  );
};

export const checkEmail = (data?: Account[], email?: string) => {
  return data?.some((user: Account) => user.email === email);
};

export const checkPassword = (data?: Account[], password?: string) => {
  return data?.some((user: Account) => user.password === password);
};

export const checkValidationStyles = (
  serverError: boolean,
  rules: FieldError | undefined,
  isDirty: boolean,
) => {
  const style = serverError || rules ? InputType.error : isDirty ? InputType.info : InputType.default;
  const theme = serverError || rules ? InputTheme.error : isDirty ? InputTheme.info : InputTheme.default

  return { style, theme };
};
