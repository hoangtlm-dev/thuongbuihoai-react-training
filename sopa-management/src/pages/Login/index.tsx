import { memo, useRef } from 'react';
import { useForm, Controller } from 'react-hook-form';
import Input, { InputTheme, InputType } from '../../components/common/Input';
import Button, { ButtonType } from '../../components/common/Button';
import Text, { SizeType } from '../../components/common/Text';
import './login.css';
import { useMutation } from 'react-query';
import axios from 'axios';
import { ACCOUNT_URL } from '../../constants/url';
import { createMemoryHistory } from 'history';
import { setStorage } from '../../helpers/storage';

interface LoginProps {
  email: string;
  password: string;
}

const Login: React.FC = () => {
  const history = createMemoryHistory();
  const emailRef = useRef<HTMLInputElement | null>(null);
  const passwordRef = useRef<HTMLInputElement | null>(null);

  const { register, handleSubmit, control } = useForm<LoginProps>();

  const loginMutation = useMutation((credentials: LoginProps) =>
    axios.post(ACCOUNT_URL, credentials)
  );

  const onSubmit = async () => {
    try {
      const email = emailRef.current?.value || '';
      const password = passwordRef.current?.value || '';
      const data = { email, password };
      const response = await loginMutation.mutateAsync(data);
      const token = response.data;
      console.log('token', token);

      setStorage('token', token);
      history.push('/');
      console.log('success');
    } catch (error: any) {
      // Handle error here
      console.log('catch');
      console.error('Login failed:', error.message);
    }
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)} className='form'>
      <Text text='Login' type={SizeType.extraMedium} />
      <div className='form-input'>
        <Controller
          name='email'
          control={control}
          render={({ fieldState: { error } }) => (
            <div className='form-email'>
              {error ? (
                <Input
                  label='Email'
                  type={error ? InputType.error : InputType.info}
                  theme={error ? InputTheme.error : InputTheme.info}
                />
              ) : (
                <Input
                  label='Email'
                  ref={(e) => {
                    register('email');
                    emailRef.current = e;
                  }}
                  type={InputType.default}
                />
              )}
            </div>
          )}
        />

        <Controller
          name='password'
          control={control}
          render={({ fieldState: { error } }) => (
            <div className='form-password'>
              {error ? (
                <Input
                  label='Password'
                  type={error ? InputType.error : InputType.info}
                  theme={error ? InputTheme.error : InputTheme.info}
                />
              ) : (
                <Input
                  label='Password'
                  ref={(e) => {
                    register('password');
                    passwordRef.current = e;
                  }}
                  type={InputType.default}
                />
              )}
            </div>
          )}
        />
        <Button children='Login' type={ButtonType.tertiary} submit='submit' />
      </div>
    </form>
  );
};

export default memo(Login);
