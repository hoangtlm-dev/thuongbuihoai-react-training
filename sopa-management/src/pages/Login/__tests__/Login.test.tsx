// Libs
import '@testing-library/jest-dom';

// Components
import Login from '..';

// Helpers
import { renderWithRouterAndQuery } from '../../../helpers/testUtils';
import { fireEvent, screen, waitFor } from '@testing-library/react';
import { checkLogin } from '@helpers/common';
import { MOCK_ACCOUNTS } from '@mocks/account';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(),
}));

jest.mock('@helpers/common', () => ({
  ...jest.requireActual('@helpers/common'),
  checkLogin: jest.fn(),
}));

describe('Login component', () => {
  test('should render Login component', () => {
    const { getByTestId } = renderWithRouterAndQuery(<Login />);
    expect(getByTestId('login')).toBeInTheDocument();
  });

  test('should not display error when value is valid', async () => {
    const mockNavigate = jest.fn();
    (checkLogin as jest.Mock).mockReturnValue({
      data: MOCK_ACCOUNTS,
      email: 'example@gmail.com',
      password: '123456'
    });

    jest
      .spyOn(require('react-router-dom'), 'useNavigate')
      .mockReturnValue(mockNavigate);

    renderWithRouterAndQuery(<Login />);

    fireEvent.input(screen.getByRole('textbox', { name: /email/i }), {
      target: {
        value: 'example@gmail.com'
      }
    });

    fireEvent.input(screen.getByRole('textbox', { name: /password/i }), {
      target: {
        value: '123456'
      }
    });

    fireEvent.submit(screen.getByRole('button'));
    expect(screen.getByRole('textbox', { name: /email/i })).toHaveValue(
      'example@gmail.com'
    );
    expect(screen.getByRole('textbox', { name: /password/i })).toHaveValue(
      '123456'
    );
    await waitFor(() => expect(mockNavigate).toHaveBeenCalledWith('/'));
  });

  test('should display error when value is invalid', async () => {
    const mockNavigate = jest.fn();
    (checkLogin as jest.Mock).mockReturnValue(false);

    jest
      .spyOn(require('react-router-dom'), 'useNavigate')
      .mockReturnValue(mockNavigate);

    renderWithRouterAndQuery(<Login />);

    fireEvent.input(screen.getByRole('textbox', { name: /email/i }), {
      target: {
        value: 'examples@gmail.com'
      }
    });

    fireEvent.input(screen.getByRole('textbox', { name: /password/i }), {
      target: {
        value: '1234567'
      }
    });

    fireEvent.submit(screen.getByRole('button'));

    await waitFor(() => expect(mockNavigate).not.toHaveBeenCalledWith('/'));
  });

  test('matches snapshot', () => {
    const { container } = renderWithRouterAndQuery(<Login />);
    expect(container).toMatchSnapshot();
  });
});
