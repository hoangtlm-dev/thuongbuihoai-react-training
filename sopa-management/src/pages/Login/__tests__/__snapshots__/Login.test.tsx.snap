// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component matches snapshot 1`] = `
<div>
  <form
    class="form"
    data-testid="login"
  >
    <p
      class=" text-extra-medium"
      data-testid="text"
    >
      Login
    </p>
    <div
      class="form-input"
    >
      <div
        class="form-email"
      >
        <div
          class="input-wrapper"
          data-testid="input-value"
        >
          <label
            class="label label-"
          >
            Email
          </label>
          <input
            aria-hidden="true"
            class="input input-"
            id=""
            name="email"
            placeholder=""
            value=""
          />
        </div>
      </div>
      <div
        class="form-password"
      >
        <div
          class="input-wrapper"
          data-testid="input-value"
        >
          <label
            class="label label-"
          >
            Password
          </label>
          <input
            aria-hidden="true"
            class="input input-"
            id=""
            name="password"
            placeholder=""
            value=""
          />
        </div>
      </div>
      <button
        aria-label="Login"
        class=" btn btn-tertiary"
        data-testid="button"
        type="submit"
      >
        Login
      </button>
    </div>
  </form>
</div>
`;
